pallets/clusters/src/lib.rs

```diff
@@ -0,0 +1,476 @@
+	#[pallet::config]
+	pub trait Config:
+		frame_system::Config
+		+ pallet_balances::Config
+		+ pallet_proxy::Config
+		+ pallet_timestamp::Config
+	{
+		/// Because this pallet emits events, it depends on the runtime's definition of an event.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+	}
+
+	/// **StorageMap** that maps a **Cluster** ID to its data.
+	#[pallet::storage]
+	pub type Clusters<T: Config> = StorageMap<_, Identity, Hash128, Cluster<T::AccountId>>;
+
+	/// **StorageMap** that maps a **AccountId** with a list of **Cluster** owned by the account.
+	#[pallet::storage]
+	pub type ClustersByOwner<T: Config> = StorageMap<_, Twox64Concat, T::AccountId, Vec<Hash128>>;
+
+	/// **StorageMap** that maps a **Role** with its **RoleSettings**.
+	#[pallet::storage]
+	pub type RoleToSettings<T: Config> = StorageMap<_, Twox64Concat, Hash128, Vec<RoleSetting>>;
+
+	#[allow(missing_docs)]
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		ClusterCreated {
+			cluster_hash: Hash128,
+		},
+		RoleCreated {
+			role_hash: Hash128,
+		},
+		RoleEdited {
+			role_hash: Hash128,
+		},
+		RoleDeleted {
+			role_hash: Hash128,
+		},
+		/// A new sponsored account was added
+		ProxyAccountAdded {
+			cluster_account: T::AccountId,
+			proxy_account: T::AccountId,
+		},
+	}
+
+	// Errors inform users that something went wrong.
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The cluster already exists.
+		ClusterExists,
+		/// Cluster not found
+		ClusterNotFound,
+		/// The role in the cluster already exists.
+		RoleExists,
+		/// Element not found
+		RoleNotFound,
+		/// Missing permission to perform an operation
+		NoPermission,
+		/// Invalid inputs
+		InvalidInputs,
+		/// Technical error not categorized
+		SystematicFailure,
+		/// The owner is not correct.
+		OwnerNotCorrect,
+		/// The member already exists in the cluster
+		MemberExists,
+		/// Member not found in the cluster
+		MemberNotFound,
+		/// Account proxy already associated with the cluster account
+		AccountProxyAlreadyExists,
+		/// Too many proxies associated with the cluster
+		TooManyProxies,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Create a **Cluster** passing a name as input.
+		#[pallet::weight(10_000 + T::DbWeight::get().writes(1).ref_time())]
+		pub fn create_cluster(origin: OriginFor<T>, name: Vec<u8>) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			ensure!(!name.len().is_zero(), Error::<T>::InvalidInputs);
+
+			let extrinsic_index = <frame_system::Pallet<T>>::extrinsic_index()
+				.ok_or(Error::<T>::SystematicFailure)?;
+
+			let cluster_id = blake2_128(
+				&[name.clone(), extrinsic_index.clone().encode(), who.clone().encode()].concat(),
+			);
+
+			sp_std::if_std! {
+				// This code is only being compiled and executed when the `std` feature is enabled.
+				println!("CLUSTER_ID: {:?}", cluster_id);
```

Comment:
println is wrong, you should use log crate as it can be useful and easily enabled/disabled via env

Reply:
oh damn, sorry! Forgot to remove it! just a quick super dirty way to see things in log when I make tests.